modules/core/src/main/scala/parser.scala

```diff
@@ -3,96 +3,98 @@
 
 package edu.gemini.grackle
 
-import atto._
-import Atto.{token => _, braces => _, parens => _, squareBrackets => _, _}
-import atto.parser.character.char
+import cats.parse.{Parser, Parser0}
+import cats.parse.Parser._
+import cats.parse.Numbers._
+import cats.parse.Rfc5234.{cr, crlf, digit, hexdig, lf}
 import cats.implicits._
 import CommentedText._
+import Literals._
 
 object GraphQLParser {
 
   def keyword(s: String) = token(string(s))
 
-  lazy val Document: Parser[Ast.Document] =
-    many(whitespace | comment) ~> many(Definition) <~ endOfInput
+  lazy val Document: Parser0[Ast.Document] =
+    many(whitespace | comment) *> many(Definition) <* Parser.end
 
-  lazy val Definition =
+  lazy val Definition: Parser[Ast.Definition] =
     ExecutableDefinition | TypeSystemDefinition // | TypeSystemExtension
 
-  lazy val TypeSystemDefinition =
-    SchemaDefinition | TypeDefinition | DirectiveDefinition
+  lazy val TypeSystemDefinition: Parser[Ast.TypeSystemDefinition] =
+    SchemaDefinition | TypeDefinition.backtrack | DirectiveDefinition
 
-  lazy val SchemaDefinition =
+  lazy val SchemaDefinition: Parser[Ast.SchemaDefinition] =
     for {
```

Comment:
replace flatMap with mapN

Reply:
I assume this is for performance reasons?

Reply:
@johnynek this actually also doesn't work as it infers `Parser0` as the result type (via `Tuple3SemigroupalOps`). I can work around it with `~` and nested tuples. Is there a nicer way?